Source/WebKitLegacy/WebCoreSupport/LegacyWebPageDebuggable.cpp

```diff
@@ -0,0 +1,124 @@
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "LegacyWebPageDebuggable.h"
+
+#include <WebCore/Document.h>
+#include <WebCore/InspectorController.h>
+#include <WebCore/Page.h>
+#include <wtf/TZoneMallocInlines.h>
+
+WTF_MAKE_TZONE_ALLOCATED_IMPL(LegacyWebPageDebuggable);
+
+LegacyWebPageDebuggable::LegacyWebPageDebuggable(LegacyWebPageInspectorController& inspectorController, WebCore::Page& page)
+    : m_inspectorController(&inspectorController)
+    , m_page(&page)
+{
+}
+
+Ref<LegacyWebPageDebuggable> LegacyWebPageDebuggable::create(LegacyWebPageInspectorController& controller, WebCore::Page& page)
+{
+    return adoptRef(*new LegacyWebPageDebuggable(controller, page));
+}
+
+String LegacyWebPageDebuggable::name() const
+{
+    String result;
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, &result] {
+        if (RefPtr page = m_page.get()) {
+            if (RefPtr document = page->localTopDocument())
+                result = document->title().isolatedCopy();
+        }
+    });
+    return result;
+}
+
+String LegacyWebPageDebuggable::url() const
+{
+    String result;
+    callOnMainThreadAndWait([this, protectedThis = Ref { *this }, &result] {
+        RefPtr page = m_page.get();
+        if (!page)
+            return;
+
+        result = page->mainFrameURL().string().isolatedCopy();
+        if (result.isEmpty())
+            result = "about:blank"_s;
+    });
+    return result;
+}
+
+bool LegacyWebPageDebuggable::hasLocalDebugger() const
+{
+    bool result;
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, &result] {
+        RefPtr controller = m_inspectorController.get();
+        result = controller && controller->hasLocalFrontend();
+    });
+    return result;
+}
+
+void LegacyWebPageDebuggable::connect(Inspector::FrontendChannel& frontendChannel, bool isAutomaticConnection, bool immediatelyPause)
+{
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, &frontendChannel, isAutomaticConnection, immediatelyPause] {
+        if (RefPtr controller = m_inspectorController.get())
+            controller->connectFrontend(frontendChannel, isAutomaticConnection, immediatelyPause);
+    });
+}
+
+void LegacyWebPageDebuggable::disconnect(Inspector::FrontendChannel& frontendChannel)
+{
+    callOnMainRunLoopAndWait([this, protectedThis = Ref { *this }, &frontendChannel] {
+        if (RefPtr controller = m_inspectorController.get())
+            controller->disconnectFrontend(frontendChannel);
+    });
+}
+
+void LegacyWebPageDebuggable::dispatchMessageFromRemote(String&& message)
+{
+    callOnMainThreadAndWait([this, protectedThis = Ref { *this }, message = WTFMove(message).isolatedCopy()]() mutable {
```

Comment:
Should this be `callOnMainThreadAndWait` or `callOnMainRunLoopAndWait`?

Reply:
I'll use `MainThread` consistent with `PageDebuggable`.